src/access/AccessControl.cpp

```diff
@@ -161,37 +161,51 @@ namespace Access {
 
 AccessControl::Entry::Delegate AccessControl::Entry::mDefaultDelegate;
 AccessControl::EntryIterator::Delegate AccessControl::EntryIterator::mDefaultDelegate;
-AccessControl::Delegate AccessControl::mDefaultDelegate;
 
-CHIP_ERROR AccessControl::Init()
+CHIP_ERROR AccessControl::Init(AccessControl::Delegate * delegate)
 {
+    VerifyOrReturnError(!mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(delegate != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
     ChipLogProgress(DataManagement, "AccessControl: initializing");
-    return mDelegate.Init();
+
+    // delegate can never be null. This was already checked
+    mDelegate         = delegate;
+    CHIP_ERROR retval = mDelegate->Init();
+    mIsInitialized    = (CHIP_NO_ERROR == retval);
+    return retval;
 }
 
 CHIP_ERROR AccessControl::Finish()
 {
+    VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
     ChipLogProgress(DataManagement, "AccessControl: finishing");
-    return mDelegate.Finish();
+    CHIP_ERROR retval = mDelegate->Finish();
```

Comment:
should we set mDelegate to nullptr here?

Reply:
Generally I wonder if mDelegate being null or not could be used as a initialized flag. They seem to always be equivalent: with it non-null  we are initialized ,with it null we can be uninitialized.

Reply:
Done